cmd/mesh/upgrade-hooks.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mesh
+
+import (
+	"strings"
+
+	"istio.io/operator/pkg/manifest"
+)
+
+type hook func(kubeClient manifest.ExecClient, istioNamespace,
+	currentVer, targetVer, currentValues, targetValues string, dryRun bool)
+
+func runPreUpgradeHooks(kubeClient manifest.ExecClient, istioNamespace,
+	currentVer, targetVer, currentValues, targetValues string, dryRun bool) {
+	for _, h := range preUpgradeHooks {
+		h(kubeClient, istioNamespace, currentVer, targetVer,
+			currentValues, targetValues, dryRun)
+	}
+}
+
+func runPostUpgradeHooks(kubeClient manifest.ExecClient, istioNamespace,
+	currentVer, targetVer, currentValues, targetValues string, dryRun bool) {
+	for _, h := range prePostUpgradeHooks {
+		h(kubeClient, istioNamespace, currentVer, targetVer,
+			currentValues, targetValues, dryRun)
+	}
+}
+
+var preUpgradeHooks = []hook{checkInitCrdJobs}
+var prePostUpgradeHooks = []hook{}
```

Comment:
We need some version concept here. Some upgrade hooks only apply to some source/target version combo, others may be always. 
The hooks should probably be a list of structs with a source and target version strings (using go-version format) and associated function. Then the code that executes the hooks should run only the ones that match the source/target combo. It would be good to show the debug output for the hooks i.e. "Checking if pre-upgrade hook <name> matches source->target versions %s-%s." Something like that.

Reply:
Agree, hook should be a standalone concept with some components below:

- Trigger condition
- Action to take

The current code is a simple func object taking currentVer, targetVer, currentValues, targetValues etc. as input (but not all parameters are used). 
I can change this into:

1. Interface with 2 funcs (e.g., Trigger() and Act()), and a hook Executor check and run all hooks.
1. A simple <current version, target version> -> [a list of hooks] map.